Encapsulate message wire encoding into a module

jkczyz opened the pull request:
Refactor message wire encoding into a separate module so clients don't have to concern themselves with type numbers and unknown types. Provide `wire::Message` and `wire::MessageType` abstractions for encapsulating this behavior. Messages implement the `wire::Encode` trait which extends `Writeable` for handling the message type. The module provides `read` and `write` convenience methods.

TheBlueMatt commented on lightning/src/ln/wire.rs:
```diff
@@ -0,0 +1,411 @@
+
+impl Message {
+	/// Returns the type that was used to decode the message payload.
+	pub fn r#type(&self) -> MessageType {
```
Hmm, never seen this syntax before (and 1.22 seems unhappy about it?)

arik-so replied:
You know, after some further deliberation, I think we may wanna try to look for other words. Raw identifiers look more awkward than, say, something like type_id or get_type or message_type for the method name.

jkczyz replied:
I'm fine with something like `type_id` or `type_num`. I'd like to avoid using `message_type` since it makes the call site look awkward as `message.message_id()`.

jkczyz replied:
And `get_type` is against the prevailing Rust conventions.

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

jkczyz replied:
Reworked all commits to use `type_id` instead.

TheBlueMatt commented on lightning/src/ln/wire.rs:
```diff
@@ -0,0 +1,411 @@
+///
+/// [`Writeable`]: ../../util/ser/trait.Writeable.html
+/// [`write`]: #method.write.html
+pub trait Encode: Writeable {
```
Hmm, can we name it something a bit more descriptive? Like just call it Message so it can be referred to as wire::Message or WireMessage?

TheBlueMatt replied:
Hmmm...So I find the current text confusing, and I dont think sticking to someone elses' style guide a compelling reason to have something that reads confusing. In general, for our own internal traits, I don't really see much reason why we should care in the slightest what some rust community members think traits should be named, at least unless we have no other criteria for determining (eg, readability).

arik-so replied:
I would disagree with Matt regarding sticking to other people's style conventions. I think it makes sense to stick to commonly agreed upon conventions such that our code becomes more easy to read and modify. Given the purpose of this library is extensibility, external readers should not be confused by names, and e. g. traits should be clearly named as verbs, like in other Rust code, too.

TheBlueMatt replied:
I don't mind it as a last-resort naming criteria, but readability should always override any style conventions (considering that the conventions are there for readability to begin with), IMO.

jkczyz replied:
Is there a project convention for naming traits?

I can think of two reasonable names in replace of `Encode`, but the readability of each depends on the context.

`Encodable` (_adjective_) reads well in trait bounds but not trait in implementations.

```rust
pub fn write<E: Encodable, W: Writer>(message: &E, buffer: &mut W) ->Result<(), ::std::io::Error> {}
impl Encodable for msgs::Ping {}
```

`Encoding` (_noun_) reads well in trait implementations but not in trait bounds.

```rust
pub fn write<E: Encoding, W: Writer>(message: &E, buffer: &mut W) ->Result<(), ::std::io::Error> {}
impl Encoding for msgs::Ping {}
```

Rust guides typically refer to traits as behaviors (verbs) rather than characteristics (adjectives). In this particular instance, we're defining a single behavior. And, as such, a verb seems more natural to me. Nouns seem more useful when talking about traits defining a set of behaviors.

@TheBlueMatt You objected to `Encode` by stating that it is confusing. Could you clarify _what_ is confusing about it? It would help in determining a suitable name.

You did mentioned readability but I think that depends on (a) the context as mentioned above and (b) the way in which you read a trait bound. For the latter, you can read the colon in a trait bound as either "is" or "has trait", each of which effects the readability differently. My feeling is when reading as "has trait" the verb usage reads fine. Perhaps there are other readings that I'm not thinking of, though.

TheBlueMatt replied:
Oops lol, I guess that wasnt' clear at all. My issue is that it is confusing as it gives no context. This is Encodeable *messages*, as we already have various encoding traits which indicate something different.

jkczyz replied:
Ah, ha! Given that the `wire` module is for encoding/decoding messages over the wire, the context within the file should be clear. If needed outside the file, a clearer name could be given using import renaming.

TheBlueMatt replied:
Right, it is, however, pub, so we should be careful to make sure it has a clear name in any context, IMO.

jkczyz replied:
Chatted offline. Keeping it as `Encode` for now but I've decoupled the trait from `Writeable`. Instead, the module's `write` function will take a message bounded by both traits. We can revisit the naming in conjunction with `Writeable` in the future.

TheBlueMatt commented on lightning/src/ln/wire.rs:
```diff
@@ -0,0 +1,411 @@
+
+/// A number identifying a message to determine how it is encoded on the wire.
+#[derive(Clone, Copy)]
+pub struct MessageType {
```
Could probably be MessageType(u16), no?

jkczyz replied:
Good point! Seems it's a simple enough type to warrant a tuple struct. I've updated it in 298a7d8.

Initially, I had given it both a number and an even/odd parity. But I didn't like how I had to expose the fields as public to match on the parity. And if they were public then they could be modified such that the number and parity did not agree. So I switch to using an `is_even` method for parity.

jkczyz commented:
Rebased

jkczyz commented:
Fixed a compilation error for Rust 1.22 where I needed to fully scope anything in an `std` module.

jkczyz commented:
Fixed remaining Rust 1.22 compilation errors.

jkczyz commented:
More Travis failures because the tests don't compile under Rust 1.22.0 (see #471 for details on older releases). Will fix these up now and re-push.

jkczyz commented:
Pushed test fixes:

- Replaced returning `Result<>` types with `()` and used `unwrap()` instead of `?`.
- Replaced `to_be_bytes` and `from_be_bytes` from `std` with custom `byte_utils` equivalent.
- Removed use `std::convert::TryInto`.
